hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/util/CachedNanoClock.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.tpc.util;
+
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNegative;
+
+/**
+ * A {@link NanoClock} that doesn't use the {@link System#nanoTime()} for every call to
+ * {@link #nanoTime()} to reduce cost, but it will cache the result of previous calls for
+ * some iterations.
+ * <p/>
+ * This class is not thread-safe.
+ */
+public class CachedNanoClock implements NanoClock {
+    private static final long START_TIME = System.nanoTime();
+
+    private final int refreshPeriod;
+    private int iteration;
+    private long value;
+
+    /**
+     * Creates a new CachedNanoClock.
+     *
+     * @param refreshPeriod the period. When set to 0, the caching is disabled.
+     * @throws IllegalArgumentException when refreshPeriod smaller than 0.
+     */
+    public CachedNanoClock(int refreshPeriod) {
+        this.refreshPeriod = checkNotNegative(refreshPeriod, "refreshPeriod");
```

Comment:
Is 0 refresh period makes sense for this class?

Reply:
Better to use the StandardNanoCLock in that case. That is what happens in Unsafe:

```
 this.nanoClock = clockRefreshInterval == 0
                    ? new StandardNanoClock()
                    : new CachedNanoClock(clockRefreshInterval);
                    ```

Reply:
Then, would it make sense to change the validation to `checkPositive` instead?

Reply:
Why? The class works fine with a zero interval.

Reply:
I thought since we have a better alternative for the zero interval, we might want to not allow this but I am resolving this